src/dataframe/dataframe.jl

```diff
@@ -516,36 +516,57 @@ end
+    u = _names(df)[selected_columns]
+    lookup = Dict{Symbol, Int}(zip(u, 1:length(u)))
+    idx = Index(lookup, u)
+
+    if length(selected_columns) == 1
+        return DataFrame(AbstractVector[_columns(df)[selected_columns[1]][row_inds]], idx, copycols=false)
+    else
         # Computing integer indices once for all columns is faster
         selected_rows = T === Bool ? findall(row_inds) : row_inds
         @static if VERSION >= v"1.4"
             if length(selected_rows) >= 1_000_000 && Threads.nthreads() > 1
+                new_columns = Vector{AbstractVector}(undef, length(selected_columns))
                 @sync for i in eachindex(new_columns)
-                    Threads.@spawn new_columns[i] = old_columns[i][selected_rows]
+                    Threads.@spawn new_columns[i] = _columns(df)[selected_columns[i]][selected_rows]
                 end
+                return DataFrame(new_columns, idx, copycols=false)
             else
-                for i in eachindex(new_columns)
-                    new_columns[i] = old_columns[i][selected_rows]
-                end
+                return DataFrame(AbstractVector[dv[selected_rows] for dv in view(_columns(df), selected_columns)], idx, copycols=false)
             end
         else
-            for i in eachindex(new_columns)
-                new_columns[i] = old_columns[i][selected_rows]
-            end
+            return DataFrame(AbstractVector[dv[selected_rows] for dv in view(_columns(df), selected_columns)], idx, copycols=false)
         end
     end
-
-    return DataFrame(new_columns, Index(_names(df)[selected_columns]), copycols=false)
 end
 
-@inline Base.getindex(df::DataFrame, row_inds::AbstractVector{T}, ::Colon) where T =
-    df[row_inds, index(df)[:]]
+@inline function Base.getindex(df::DataFrame, row_inds::AbstractVector{T}, ::Colon) where T
+    @boundscheck if !checkindex(Bool, axes(df, 1), row_inds)
+        throw(BoundsError(df, (row_inds, :)))
+    end
+
+    if ncol(df) == 1
+        return DataFrame(AbstractVector[_columns(df)[1][row_inds]], copy(index(df)), copycols=false)
+    else
+        # Computing integer indices once for all columns is faster
+        selected_rows = T === Bool ? findall(row_inds) : row_inds
+        @static if VERSION >= v"1.4"
+            if length(selected_rows) >= 1_000_000 && Threads.nthreads() > 1
+                new_columns = Vector{AbstractVector}(undef, ncol(df))
+                @sync for i in eachindex(new_columns)
+                    Threads.@spawn new_columns[i] = _columns(df)[i][selected_rows]
+                end
+                return DataFrame(new_columns, copy(index(df)), copycols=false)
+            else
+                return DataFrame(AbstractVector[dv[selected_rows] for dv in _columns(df)], copy(index(df)), copycols=false)
+            end
+        else
+            return DataFrame(AbstractVector[dv[selected_rows] for dv in _columns(df)], copy(index(df)), copycols=false)
+        end
+    end
```

Comment:
Isn't there a way to avoid code duplication with the method above?

Reply:
So initially in the PR I did `index(df)[:]` and delegated to the method above. However this lead to performance degradation and I thought that `:` as column selector is common enough that we should make it fast. The difference is not big (this is the current PR):
```
julia> df = DataFrame(rand(1, 10), :auto);

julia> @benchmark $df[1:1, :]
BenchmarkTools.Trial: 
  memory estimate:  1.86 KiB
  allocs estimate:  17
  --------------
  minimum time:     1.514 μs (0.00% GC)
  median time:      1.570 μs (0.00% GC)
  mean time:        1.875 μs (3.52% GC)
  maximum time:     177.066 μs (98.59% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> @benchmark $df[1:1, 1:10]
BenchmarkTools.Trial: 
  memory estimate:  1.86 KiB
  allocs estimate:  17
  --------------
  minimum time:     1.621 μs (0.00% GC)
  median time:      1.677 μs (0.00% GC)
  mean time:        1.980 μs (3.36% GC)
  maximum time:     175.479 μs (98.61% GC)
  --------------
  samples:          10000
  evals/sample:     10
```
but I was observing it when I was doing the benchmarks (reported below in the main thread) against 0.22.5. In general the current code was hand-crafted to make sure that if user operates on small tables all is fast (I thought this case is relevant for split-apply-combine used with custom function getting a data frame as an input).

Reply:
OK I was thinking about sharing some of the code in a helper function. But maybe that's not worth it.

Reply:
I have considered this, but there are some small differences in many places (although the high level logic is the same).